bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -3259,6 +3276,69 @@ void ReplicaImp::onMessage<SimpleAckMsg>(SimpleAckMsg *msg) {
+  if (!seq_num_to_stop_at.has_value() || (msg->seqNum() != seq_num_to_stop_at.value())) {
+    LOG_ERROR(GL,
+              "Recieved invalid ReplicaRestartReadyMsg from sender_id "
+                  << std::to_string(msg->idOfGeneratedReplica()) << " with seq_num" << std::to_string(msg->seqNum()));
+    delete msg;
+    return;  // this is to handle replay attack
+  }
+  LOG_INFO(GL,
+           "Recieved ReplicaRestartReadyMsg from sender_id " << std::to_string(msg->idOfGeneratedReplica())
+                                                             << " with seq_num" << std::to_string(msg->seqNum()));
+  if (restart_ready_msgs_.find(msg->idOfGeneratedReplica()) == restart_ready_msgs_.end()) {
+    restart_ready_msgs_[msg->idOfGeneratedReplica()] = std::make_unique<ReplicaRestartReadyMsg>(msg);
+    metric_received_restart_ready_.Get().Inc();
+  } else {
+    LOG_ERROR(GL,
+              "Recieved multiple ReplicaRestartReadyMsg from sender_id "
+                  << std::to_string(msg->idOfGeneratedReplica()) << " with seq_num" << std::to_string(msg->seqNum()));
+    delete msg;
+  }
+  bool bft = bftEngine::ControlStateManager::instance().getRestartBftFlag();
+  uint32_t targetNumOfMsgs = (bft ? (config_.getnumReplicas() - config_.getfVal()) : config_.getnumReplicas());
+  if (restart_ready_msgs_.size() == targetNumOfMsgs) {
+    LOG_INFO(GL, "Target number = " << targetNumOfMsgs << " of restart ready msgs are recieved. Send resatrt proof");
+    sendReplicasRestartReadyProof();
+    ControlStateManager::instance().onRestartProof();
+  }
+}
+
+template <>
+void ReplicaImp::onMessage<ReplicasRestartReadyProofMsg>(ReplicasRestartReadyProofMsg *msg) {
+  auto seq_num_to_stop_at = ControlStateManager::instance().getCheckpointToStopAt();
+  if (!seq_num_to_stop_at.has_value() || (msg->seqNum() != seq_num_to_stop_at.value())) {
+    LOG_ERROR(GL,
+              "Recieved invalid ReplicasRestartReadyProofMsg from sender_id "
+                  << std::to_string(msg->idOfGeneratedReplica()) << " with seq_num" << std::to_string(msg->seqNum()));
+    delete msg;
+    return;  // this is to handle replay attack
+  }
+  LOG_INFO(GL,
+           "Recieved  ReplicasRestartReadyProofMsg from sender_id "
+               << std::to_string(msg->idOfGeneratedReplica()) << " with seq_num" << std::to_string(msg->seqNum()));
+  metric_received_restart_proof_.Get().Inc();
+  ControlStateManager::instance().onRestartProof();
+  delete msg;
+}
+
+void ReplicaImp::sendReplicasRestartReadyProof() {
+  auto seq_num_to_stop_at = ControlStateManager::instance().getCheckpointToStopAt();
+  if (seq_num_to_stop_at.has_value()) {
+    unique_ptr<ReplicasRestartReadyProofMsg> restartProofMsg(
+        ReplicasRestartReadyProofMsg::create(config_.getreplicaId(), seq_num_to_stop_at.value()));
+    for (auto &[_, v] : restart_ready_msgs_) {
+      (void)_;  // unused variable hack
+      restartProofMsg->addElement(v);
```

Comment:
are you signing on to this message?
any receiver will start upon receiving a single restart proof, right?

Reply:
Yes these messages are signed by the sender
Yes, replica restart on receiving single restart proof.